airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/streams.py

```diff
@@ -0,0 +1,486 @@
+        super().__init__(*args, **kwargs)
+
+        # add the custom value for generation of a zendesk domain
+        self._subdomain = subdomain
+
+    @property
+    def url_base(self) -> str:
+        return f"https://{self._subdomain}.zendesk.com/api/v2/"
+
+    @staticmethod
+    def _parse_next_page_number(response: requests.Response) -> Optional[int]:
+        """Parses a response and tries to find next page number"""
+        next_page = response.json()["next_page"]
+        if next_page:
+            return dict(parse_qsl(urlparse(next_page).query)).get("page")
+        return None
+
+    def backoff_time(self, response: requests.Response):
+        """
+        The rate limit is 700 requests per minute
+        # monitoring-your-request-activity
+        See https://developer.zendesk.com/api-reference/ticketing/account-configuration/usage_limits/
+        The response has a Retry-After header that tells you for how many seconds to wait before retrying.
+        """
+        retry_after = response.headers.get("Retry-After")
+        if retry_after:
+            return int(retry_after)
+        # the header X-Rate-Limit returns a amount of requests per minute
+        # we try to wait twice as long
+        rate_limit = float(response.headers.get("X-Rate-Limit") or 0)
+        if rate_limit:
+            return (60.0 / rate_limit) * 2
+        # default value if there is not any headers
+        return 60
+
+    @staticmethod
+    def str2datetime(s):
+        """convert string to datetime object"""
+        if not s:
+            return None
+        return datetime.strptime(s, DATETIME_FORMAT)
+
+    @staticmethod
+    def datetime2str(dt):
+        """convert datetime object to string"""
+        return datetime.strftime(dt.replace(tzinfo=pytz.UTC), DATETIME_FORMAT)
+
+
+class UserSettingsStream(SourceZendeskSupportStream):
+    """Stream for checking of a request token and permissions"""
+
+    def path(self, *args, **kwargs) -> str:
+        return "account/settings.json"
+
+    def next_page_token(self, *args, **kwargs) -> Optional[Mapping[str, Any]]:
+        # this data without listing
+        return None
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        """returns data from API"""
+        yield from [response.json().get("settings") or {}]
+
+    def get_settings(self) -> Tuple[Mapping[str, Any], Union[str, None]]:
+        for resp in self.read_records(SyncMode.full_refresh):
+            return resp
+        raise SourceZendeskException("not found settings")
+
+
+class IncrementalBasicEntityStream(SourceZendeskSupportStream, ABC):
+    """basic stream for endpoints where an entity name can be used in a path value
+    https://<subdomain>.zendesk.com/api/v2/<entity_name>.json
+    """
+
+    # max size of one data chunk. 100 is limitation of ZenDesk
+    state_checkpoint_interval = SourceZendeskSupportStream.page_size
```

Comment:
There are streams returning records in descending order of the cursor field, which makes me think this would prematurely save state and cause data loss.

See [the docs](https://github.com/airbytehq/airbyte/blob/master/docs/connector-development/cdk-python/incremental-stream.md) for more details

Reply:
Thank you for details. You are right to the Zendesk provides only one export endpoint in ascending order. And it will be able to be a potential bug in the future.
Fixed